File: examples/C/Tiles_optimization_example.c
#include <stdio.h>
int main(void)
{ 
    int ny = 10000;
    int nx = 10000;
    int idxA, idxB;
    int a[nx*ny], b[nx*ny];

// Structured data region
    #pragma acc data copyout(a, b)
    {
        #pragma acc parallel loop present(a, b)
        for(int i=0; i<nx*ny; ++i)
            a[i] = i;
           
        #pragma acc parallel loop present(a, b) tile(32,32)
        for (int i=0; i<nx; ++i){
            for (int j=0; j<ny; ++j){
                idxB = j + (i)*nx;
                idxA = i + (j)*ny;
                b[idxB] = a[idxA];
            }
        }
    }
// End of structured data region
/*
    fprintf(stderr,"A: \n");
    for(int i=0; i<nx*ny; ++i) fprintf(stderr,"%d\n",a[i]);
    fprintf(stderr,"\n");
    fprintf(stderr,"B: \n");
    for(int i=0; i<nx*ny; ++i) fprintf(stderr,"%d\n",b[i]);
*/    
}

File: examples/Fortran/Tiles_optimization_example.f90
program tiles_optimization
    use ISO_FORTRAN_ENV, only : INT32, REAL64
    implicit none

    integer(kind=INT32 ), parameter        :: nx = 10000, ny = 10000 ! put nx, ny = 20000 on V100-32
    real   (kind=REAL64), dimension(nx,ny) :: A
    real   (kind=REAL64), dimension(ny,nx) :: B
    integer(kind=INT32 )                   :: i, j

    !$acc data copyout(A(:,:), B(:,:))
    
    !$acc parallel loop
    do j = 1, ny
        do i = 1, nx
            A(i,j) = i + (j-1)*ny
        enddo
    enddo    

    !$acc parallel loop tile(32,32)
    do j = 1, nx
        do i = 1, ny 
            B(i,j) = A(j,i)
        enddo
    enddo
    !$acc end data

    write(0,'(a2)') 'A:'
    do i = 1, 50
        write(0,"(5f20.8)") A(1:5,i)
    enddo
   
    write(0,'(a2)') 'B:'
    do i = 1, 50
        write(0,"(5f20.8)") B(1:5,i)
    enddo

end program tiles_optimization      
